content/blog/2018/delayering-istio/delayering-istio.md

```diff
@@ -0,0 +1,216 @@
+
+## AppSwitch
+
+Not unlike the container runtime, AppSwitch consists of a client and a daemon that speak REST with each other.  Both client and the daemon are built as one self-contained binary, `ax`.  Client transparently plugs into the application and tracks its system calls related to network connectivity and notifies the daemon about their occurrences.  As an example, let’s say an application makes the `connect(2)` system call to the service IP of a Kubernetes service.  AppSwitch client intercepts the connect, nullifies it and notifies the daemon about its occurrence along with some context that includes the system call arguments.  Daemon would then handle the system call, perhaps by directly connecting to the Pod IP of the upstream server on behalf of the application.
+
+It is important to note that no data is forwarded between AppSwitch client and daemon.  They are designed to exchange file descriptors (FDs) over a Unix socket to avoid having to copy data.  Note also that client is not a separate process.  Rather it directly runs in the context of the application itself.  There is no data copy between the application and AppSwitch client either.
+
+## Delayerification
+
+Now that we have an idea about what AppSwitch does, let’s look at the layers that it optimizes away from a standard service mesh.
+
+### Network Devirtualization
+
+Kubernetes offers simple and well-defined network constructs to the microservice applications it runs.  In order to support them however, it imposes specific [requirements](https://kubernetes.io/docs/concepts/cluster-administration/networking/) on the underlying network.  Meeting those requirements is often not easy.  The goto solution of adding another layer is typically adopted to satisfy the requirements.  In most cases the additional layer consists of a network overlay that sits between Kubernetes and underlying network.  Traffic produced by the applications is encapsulated at the source and decapsulated at the target which not only costs network resources but also takes up compute cores due to the additional processing.
+
+Because AppSwitch arbitrates what the application sees through its touchpoints with the platform, it projects a consistent virtual view of the underlying network to the application similar to an overlay but without introducing an additional layer of processing along the datapath.  Just to draw a parallel to containers, the inside of a container looks and feels like a VM.  However the underlying implementation does not intervene along the high-incidence control paths of low-level interrupts etc.
+
+AppSwitch can be injected into a standard Kubernetes manifest (similar to Istio injection) such that the application’s network is directly handled by AppSwitch bypassing any network overlay underneath.  More details to follow in just a bit.
```

Comment:
application's network traffic are

Reply:
Again, trying to avoid giving an impression that AppSwitch handles network traffic.